### PR TITLE
Add script to generate `docker-compose.yaml` for Qdrant cluster

### PR DESCRIPTION
**Pull request by ffuugoo**
_Edited by @timvisee:_

This adds a script to generate a Docker compose configuration to easily start a small cluster in Docker.

I've been using this a few times, so I decided to make it ready to merge into our main repo.

Example usage:

```bash
./tools/generate_docker_compose_cluster.sh 5 > docker-compose.yaml

docker-compose up
```

### CODE REVIEW

**Comment by ffuugoo** [comment]
I've cancelled all workflows, cause there is no reason to run them now. ¯\\\_(ツ)\_/¯

**Review by timvisee — APPROVED** [review]

**Comment by ffuugoo** [comment]
@timvisee note there's also a small code change in this PR (though, it's a mostly useful one).

**Comment by timvisee** [comment]
> @timvisee note there's also a small code change in this PR (though, it's a mostly useful one).

I noticed. I didn't think there was a need to remove it. :smile:

File: lib/collection/src/operations/types.rs [hunk 1]
```diff
@@ -584,7 +585,17 @@ impl From<RequestError<tonic::Status>> for CollectionError {
...
             RequestError::FromClosure(status) => status.into(),
-            RequestError::Tonic(err) => CollectionError::service_error(format!("{err}")),
+            RequestError::Tonic(err) => {
+                let mut msg = err.to_string();
+                let mut err = err.source();
+
+                while let Some(src) = err {
+                    msg += &format!(": {src}");
+                    err = src.source();
+                }
```
Comment (ffuugoo) [inline lib/collection/src/operations/types.rs:595]: ...and use it to build `msg` with fewer allocations.

```suggestion
                while let Some(src) = err {
                    err = src.source();
                    write!(&mut msg, ": {src}").unwrap(); // Writing into `String` never fails
                }
```

**Review by ffuugoo — COMMENTED** [review]
🚲

File: lib/collection/src/operations/types.rs (hunk 1 above)
Comment (timvisee) [reply lib/collection/src/operations/types.rs:595]: Nice.

I did attempt a quick rewrite with the [`successors`](https://doc.rust-lang.org/std/iter/fn.successors.html) iterator as well, but lifetimes make it hard.

Comment (ffuugoo) [reply lib/collection/src/operations/types.rs]: > I did attempt a quick rewrite with the [successors](https://doc.rust-lang.org/std/iter/fn.successors.html) iterator as well, but lifetimes make it hard.

Awesome find! Works this way:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cf3463710648c5cd21b259f5e7470a04

Comment (timvisee) [reply lib/collection/src/operations/types.rs]: The types are a bit different, which makes it harder:

image

I wanted to use a `successors().map().collect().join(": ")` to make it a bit nicer. But whatever, Rust doesn't let me. :smiling_face_with_tear:

Comment (ffuugoo) [reply lib/collection/src/operations/types.rs]: Works like this then (notice `|&src|` in closure parameters):

```rust
for src in iter::successors(err.source(), |&src| src.source()) {
    write!(&mut msg, ": {src}").unwrap(); // Writing into `String` never fails
}
```

Comment (timvisee) [reply lib/collection/src/operations/types.rs]: Ugh, not sure how I missed the `&`. Thanks!

We can also do 😆:

```rust
let msg = std::iter::once(err.to_string())
    .chain(successors(err.source(), |&src| src.source()).map(ToString::to_string))
    .collect::<Vec<_>>()
    .join(": ");
CollectionError::service_error(msg)
```

**Pull request merged**